working/0508 - strict-inference/strict-inference.md

```diff
@@ -0,0 +1,350 @@
+  var b = {};   // Inference failure
+  final c = []; // Inference failure
+  const d = []; // Inference failure
+
+  void mapFunction(map = {}) {}             // Inference failure
+  var d = args.isEmpty ? [] : args.take(1); // Inference failure
+  var e = args ?? [];                       // OK; the type of the right side of
+                                            // `??` is inferred from the left.
+  dynamic returnsList() => [];              // Inference failure
+
+  int len(List list) => list.length;
+  List h = []; // OK; `List h` is shorthand for `List<dynamic> h`.
+  len([]);     // OK; `List list` is shorthand for `List<dynamic> list`.
+}
+```
+
+The failure is reported on the collection literal.
+
+### Instance creation
+
+Instantiating a generic class without explicit type argument(s), in which one
+or more type arguments cannot be inferred from downwards or upwards inference
+is considered an inference failure. This includes type parameters with an
+explicit bound.
+
+```dart
+class C<T> {
+  T t;
+
+  C();
+  C.of(this.t);
+}
+
+class D<T extends num> {}
+
+void main() {
+  var f = Future.value();          // Inference failure
+  var g = Future.error("Error");   // Inference failure
+  Future<void> h = Future.value(); // OK
+  var i = Future<void>.value();    // OK
+  var j = Future.value(7);         // OK
+  var l = List();                  // Inference failure
+  var c1 = C();                    // Inference failure
+  var d = D();                     // Inference failure
+  C<int> c2 = C();                 // OK
+  C c3 = C<int>();                 // OK
+  var c4 = C.of(42 as dynamic);    // OK
+}
+
+Future<void> fn() => Future.error("Error"); // OK
+```
+
+The failure is reported on the constructor call. Only one failure is reported,
+even if multiple type arguments cannot be inferred, as the failure is on the
+type of object being instantiated.
+
+### Function call
+
+Calling a generic function without explicit type arguments, such that one or
+more type arguments cannot be inferred from downwards or upwards inference is
+considered an inference failure. This includes type parameters with an explicit
+bound.
+
+```dart
+T f1<T>(dynamic a) => a as T;
+
+void main() {
+  f1(7);                    // Inference failure
+  var a = f1(7);            // Inference failure
+  var b = [1, 2, 3].cast(); // Inference failure
+}
+```
+
+The inference failure is reported on the function call.
+
+### Function return types
+
+Declaring a recursive local function, a top-level function, a method, or a
+function-typed function parameter without a return type is an inference
+failure. The return type of non-recursive local functions can always be
+inferred from downwards or upwards inference, as it will have a body, and the
+return type of the body is known (even if there are inference failures within).
+
+TODO(srawlins): Examples
+
+## Cascading failures
+
+Cascading inference failures are not reported. That is, when inference cannot
```

Comment:
It's probably sufficient, and perhaps better, to simply say that this proposal does not change the way type inference works, only what is or is not reported to the user.

Reply:
I've clarified this at the top, good  suggestion.

I think I would like this section at the bottom, because these questions come up a lot. Originally Jenny even implemented strict-inference to cascade things like `var a; var b = [a];`  so that `b` also has an inference failure. The return-type example came up in internal discussions.

I made up the for-each and field-initializing parameter examples while writing the rest of the spec, so I don't feel strongly about them. But they serve as interesting examples of "there is no type here, and yet this never means an inference failure."

WDYT?

Reply:
I think it's fine if you want to leave the section, but it might be good to clarify that all of the things mentioned in this section fall out naturally from the fact that type inference has inferred `dynamic` and that there is no additional requirements being added by this section. (Basically the normative / non-normative distinction from the language spec.)

Reply:
OK, I've added that distinction, and clarified the text.